subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -51,56 +74,76 @@ public HierarchicalFileWatcherUpdater(FileWatcher watcher) {
             .map(File::toPath)
             .map(Path::toAbsolutePath)
             .collect(Collectors.toSet());
-        projectRootDirectories.clear();
-        for (Path rootPathToWatch : WatchRootUtil.resolveRootsToWatch(rootPaths)) {
-            projectRootDirectories.put(rootPathToWatch, rootPathToWatch.toString() + File.separator);
-        }
-        LOGGER.info("Now considering {} as root directories to watch", projectRootDirectories.keySet());
+        Set<Path> newRootProjectDirectories = WatchRootUtil.resolveRootsToWatch(rootPaths);
+        LOGGER.info("Now considering {} as root directories to watch", newRootProjectDirectories);
+
+        knownRootProjectDirectoriesFromCurrentBuild.clear();
+        newRootProjectDirectories.stream()
+            .map(RootProjectDirectory::new)
+            .forEach(knownRootProjectDirectoriesFromCurrentBuild::add);
+        watchedRootProjectDirectoriesFromPreviousBuild.removeAll(knownRootProjectDirectoriesFromCurrentBuild);
+
         updateWatchedDirectories();
     }
 
     private void updateWatchedDirectories() {
         Set<Path> directoriesToWatch = new HashSet<>();
-        shouldWatchDirectories.elementSet().forEach(shouldWatchDirectory -> {
+        trackedDirectoriesForSnapshot.values().forEach(shouldWatchDirectory -> {
             String shouldWatchDirectoryPathString = shouldWatchDirectory.toString();
-            for (Map.Entry<Path, String> entry : projectRootDirectories.entrySet()) {
-                Path projectRootDirectory = entry.getKey();
-                String projectRootDirectoryPrefix = entry.getValue();
-                if (shouldWatchDirectoryPathString.startsWith(projectRootDirectoryPrefix)) {
-                    directoriesToWatch.add(projectRootDirectory);
-                    return;
-                }
+            if (maybeWatchRootProjectDirectory(directoriesToWatch, shouldWatchDirectoryPathString, knownRootProjectDirectoriesFromCurrentBuild)) {
+                return;
+            }
+            if (maybeWatchRootProjectDirectory(directoriesToWatch, shouldWatchDirectoryPathString, watchedRootProjectDirectoriesFromPreviousBuild)) {
+                return;
             }
             directoriesToWatch.add(shouldWatchDirectory);
         });
 
         updateWatchedDirectories(WatchRootUtil.resolveRootsToWatch(directoriesToWatch));
     }
 
+    private static boolean maybeWatchRootProjectDirectory(Set<Path> directoriesToWatch, String shouldWatchDirectoryPathString, Set<RootProjectDirectory> rootProjectDirectories) {
+        for (RootProjectDirectory rootProjectDirectory : rootProjectDirectories) {
+            if (rootProjectDirectory.contains(shouldWatchDirectoryPathString)) {
+                directoriesToWatch.add(rootProjectDirectory.path);
+                return true;
+            }
+        }
+        return false;
+    }
+
     private void updateWatchedDirectories(Set<Path> newWatchRoots) {
```

Comment:
We are mutating the parameter collection, which is not nice. Making a copy of it would allow the copy to be named `watchRootsToAdd`, which would capture the symmetry with `watchRootsToRemove`.

Also, we now have `directory` (a single level of directory), `hierarchy` (a directory and all its descendants) and `watchRoot` (same), and `rootProjectDirectory` (a hierarchy at the root of a partial(?) build).

WDYT about `hierarchyToStartWatching` / `hierarchyToStopWatching`? I think we should replace all `watchRoot`s with the word `hierarchy`. This would disambiguate it from `rootProjectDirectory`.

I think this method should be called `updateWatchedHierarchies()`.